src/cli_plugin/install/zip.js

```diff
@@ -1,98 +1,120 @@
- * @param {integer} strip - Number of nested directories within the archive
- *  that should be ignored when determining the target path of an archived
- *  file.
- * @param {function} filter - A function that accepts a single parameter 'file'
- *  and returns true if the file should be extracted from the archive
- */
-export async function extractFiles(zipPath, targetPath, strip, filter) {
-  await new Promise((resolve, reject) => {
-    const unzipper = new DecompressZip(zipPath);
+        if (match) {
+          zipfile.openReadStream(entry, function (err, readable) {
+            const chunks = [];
+
+            if (err) {
+              return reject(`unable to read ${entry.fileName} within zip`);
+            }
 
-    unzipper.on('error', reject);
+            readable.on('data', chunk => chunks.push(chunk));
 
-    const options = {
-      path: targetPath,
-      strip: strip
-    };
-    if (filter) {
-      options.filter = extractFilter(filter);
-    }
+            readable.on('end', function () {
+              const contents = Buffer.concat(chunks).toString();
+              const pkg = JSON.parse(contents);
 
-    unzipper.extract(options);
+              plugins.push(Object.assign(pkg, {
+                archivePath: match[1],
+                archive: archive,
 
-    unzipper.on('extract', resolve);
+                // Plugins must specify their version, and by default that version should match
+                // the version of kibana down to the patch level. If these two versions need
+                // to diverge, they can specify a kibana.version to indicate the version of
+                // kibana the plugin is intended to work with.
+                kibanaVersion: get(pkg, 'kibana.version', pkg.version)
+              }));
+
+              zipfile.readEntry();
+            });
+          });
+        } else {
+          zipfile.readEntry();
+        }
+      });
+
+      zipfile.on('close', () => {
+        resolve(plugins);
+      });
+    });
   });
 }
 
-/**
- * Returns all files within an archive
- * @param {string} zipPath - file path to a zip archive
- * @returns {array} all files within an archive with their relative paths
- */
-export async function listFiles(zipPath) {
-  return await new Promise((resolve, reject) => {
-    const unzipper = new DecompressZip(zipPath);
+export function extractArchive(archive, targetDir, extractPath) {
+  return new Promise((resolve, reject) => {
+    yauzl.open(archive, { lazyEntries: true }, function (err, zipfile) {
+      if (err) {
+        return reject(`unable to read ${archive}`);
+      }
 
-    unzipper.on('error', reject);
+      zipfile.readEntry();
+      zipfile.on('close', resolve);
+      zipfile.on('entry', function (entry) {
+        let fileName = entry.fileName;
 
-    unzipper.on('list', (files) => {
-      files = files.map((file) => file.replace(/\\/g, '/'));
-      resolve(files);
-    });
+        if (extractPath && fileName.startsWith(extractPath)) {
```

Comment:
this properly makes the kibana dir as top level, but includes other top level dirs in the kibana directory

```
$ ls plugins/x-pack
LICENSE.txt  NOTICE.txt  elasticsearch  index.js  logstash  node_modules  package.json plugins  server
```